### PR TITLE
Kimberley Mackenzie - Carets

### PR DESCRIPTION
**Pull request by kimberleyamackenzie**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| How did you go about exploring the Edamam API, how did you try querying the API? | I read through the docs, and did a lot of trial and error in postman (a little) and my console. |
| Describe your API Wrapper.  How did you decide on the methods you created? | I have two methods- one for returning a search object (that contains all the recipes matching a query) and a method that returns one recipe object that reflects a single recipe accessed through its uri.  I decided that made the most sense in accessing those two different pieces of information.  |
| Describe an edge case or failure case test you wrote for your API Wrapper. | Ensuring that a response is generated, even if the test parameters don't have a matching recipe in the API. |
| Explain how VCR aids in testing an API. | Instead of calling to the API multiple times and being dependent on the API working/the internet connection/maxing out the calls, it records information from a single call and allows you to replay it for multiple different tests. |
| What is the Heroku URL of your deployed application? | https://recipe-finder-kimberley.herokuapp.com/  |

### CODE REVIEW

File: app/views/search/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<section class="root-page">
+  <div class="search-form">
+  <%= form_tag search_results_path, method: "get" do %>
+    <%= text_field_tag(:ingredients, 'What\'s in the fridge?', options = {:autofocus => true }) %>
```
Comment (CheezItMan) [inline app/views/search/index.html.erb:4]: I would use something like:  
`placeholder: "what\'s in the fridge"`

In the HTML options for the tag instead of setting the form field value.  It's just annoying otherwise.

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        <span>></span>
+      </button>
+      <ul class="nav">
+        <li class="search"><%= link_to "New Search", root_path %></a></li>
+        <li class="credit"><a href="https://www.edamam.com/">Powered by Edamam</a></li>
+      </ul>
+    </header>
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script>
+
+    <script type="text/javascript">
```
Comment (CheezItMan) [inline app/views/layouts/application.html.erb:27]: Ohhh JavaScript!

File: app/views/search/create.html.erb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+<section class="recipe-list">
+  <% if @all_recipes.count == 0 %>
+  <h1 class="none">No recipes available<br />
+  <%= link_to "New Search", root_path %></h1>
+  <% end %>
+
+  <% @all_recipes.each do |recipe| %>
+  <div class="one-recipe row">
+    <div class= "search-image small 5-columns large 3-columns"> <%= image_tag(recipe.image) %> </div>
```
Comment (CheezItMan) [inline app/views/search/create.html.erb:9]: If you use an item icon like this image, make it a link, it's annoying to click on a picture and have nothing happen.

File: test/controllers/search_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+describe SearchController do
```
Comment (CheezItMan) [inline test/controllers/search_controller_test.rb]: 🙅‍♂️

File: test/lib/search_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  it "succeeds for the create action to generate search results for longer query" do
+    VCR.use_cassette("list_all") do
+      get search_results_path("chicken, lemon, balsamic")
+      must_respond_with :success
+    end
+  end
+
+  it "succeeds in showing information about a single recipe" do
+    VCR.use_cassette("list_another_one") do
+      # get search_results_path("lemon")
```
Comment (CheezItMan) [inline test/lib/search_controller_test.rb]: You do need to test show actions.

You also need to test negative cases (no search results and bad URIs for the show action).

File: test/lib/search_test.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
+require 'test_helper'
+
+describe Search do
+  before do
+    VCR.use_cassette("list_all") do
```
Comment (CheezItMan) [inline test/lib/search_test.rb:5]: I would also test required initial fields in `initialize`

File: test/test_helper.rb [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+VCR.configure do |config|
+  config.cassette_library_dir = 'test/cassettes' # folder where casettes will be located
+  config.hook_into :webmock # tie into this other tool called webmock
+  config.default_cassette_options = {
+    :record => :new_episodes,    # record new data when we don't have it yet
+    :match_requests_on => [:method, :uri, :body] # The http method, URI and body of a request all need to match
+  }
+  # Don't leave our Slack token lying around in a cassette file.
+  config.filter_sensitive_data("<EDAMAM_KEY>") do
+    ENV['EDAMAM_KEY']
```
Comment (CheezItMan) [inline test/test_helper.rb:20]: The ID should also be filtered!

File: test/lib/edamam_api_wrapper_test.rb [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+      search.must_be_instance_of Search
+      search.hits.count.must_equal 0
+    end
+  end
+
+  it "Can find one specific recipe using the URI" do
+    VCR.use_cassette("list_one") do
+      search = EdamamApiWrapper.show_recipe(EdamamApiWrapper.find_recipes("lemon").hits[0]['recipe']['uri'])
+      search.must_be_instance_of HTTParty::Response
```
Comment (CheezItMan) [inline test/lib/edamam_api_wrapper_test.rb]: negative case???

**Comment by CheezItMan** [comment]
# API Muncher
## What We're Looking For

| Feature | Feedback    |
| :------------- | :------------- |
| **Core Requirements** |   |
| Git hygiene | Good number of commits and good commit messages |
| Comprehension questions	| Check |
| **General** |  |
| Rails fundamentals (RESTful routing, use of named paths) | Check |
| Semantic HTML | Check |
| Errors are reported to the user |   Yes with 0 search results, but the site crashes for a bad item URI.|
| API Wrapper to handle the API requests | Check |
| Controller testing | **Gah**  No controller tests in the controller folder!  What the heck!  In the `lib` folder you're testing the controller, but not the show action or negative cases.  |
| Lib testing | Check, but no negative case testing |
| Search Functionality | Check |
| List Functionality | Check |
| Show individual item functionality (link to original recipe opens in new tab) | Check |
| **Styling** |  |
| Foundation Styling for responsive layout | The show individual recipe page looks a little weird at medium size.  |
| List View shows 10 items at a time/pagination | Check |
| The app is styled to create an attractive user interface | Check|
| **API Features** | |
| The App attributes Edaman |Check |
| The VCR casettes do not contain the API key | The key yes, the ID no |
| **External Resources** |  |
| Link to deployed app on Heroku | Check |
|  **Overall** | Pretty good, but it's clear you need more practice with testing.  I left some notes in your code. |

File: test/controllers/search_controller_test.rb [hunk 9]
```diff
@@ -0,0 +1,57 @@
+require 'test_helper'
```
Comment (CheezItMan) [inline test/controllers/search_controller_test.rb:1]: This is really good, nice work!
